File: scripts/crawl_gwangju.ts
import * as cheerio from "cheerio";
import { createClient } from "@supabase/supabase-js";
import * as dotenv from "dotenv";

dotenv.config({ path: ".env.local" });

// Supabase 설정 (bycmqxndlyiocnnsecxb)
const SUPABASE_URL = process.env.SUPABASE_URL || "https://bycmqxndlyiocnnsecxb.supabase.co";
const SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_KEY || "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImJ5Y21xeG5kbHlpb2NubnNlY3hiIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3MDY1Nzk5OCwiZXhwIjoyMDg2MjMzOTk4fQ.qOr1RWbdJCfPE8c29O4hKPkFtNuOmyuVoI0oU_9LSIc";

const supabase = createClient(SUPABASE_URL, SERVICE_ROLE_KEY, {
  auth: {
    autoRefreshToken: false,
    persistSession: false,
    detectSessionInUrl: false
  }
});

// 광주광역시청 보도자료 URL
const LIST_URL = "https://www.gwangju.go.kr/boardList.do?pageId=www789&boardId=BD_0000000027";
const DETAIL_URL_BASE = "https://www.gwangju.go.kr/boardView.do?pageId=www789&boardId=BD_0000000027&seq=";

const HEADERS = {
  "User-Agent":
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
};

async function fetchHtml(url: string) {
  try {
    const res = await fetch(url, { headers: HEADERS });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    return await res.text();
  } catch (error) {
    console.error(`Failed to fetch ${url}:`, error);
    return null;
  }
}

async function crawlGwangju() {
  console.log("Fetching list...");
  const listHtml = await fetchHtml(LIST_URL);
  
  if (!listHtml) return;

  const $ = cheerio.load(listHtml);
  
  // div.body_row 태그 확인 (반응형 게시판 구조)
  const $rows = $(".board_list_body .body_row");
  console.log(`Found ${$rows.length} rows. Processing top 5...`);

  // 상위 5개만 처리
  for (let i = 0; i < Math.min(5, $rows.length); i++) {
    const row = $rows[i];
    const $row = $(row);

    try {
      // 제목 셀 찾기 (.subject)
      const $subjectDiv = $row.find(".subject");
      if ($subjectDiv.length === 0) continue;

      const $link = $subjectDiv.find("a");
      const title = $link.text().trim();
      const href = $link.attr("href");

      // href에서 seq 추출
      const seqMatch = href?.match(/seq=(\d+)/);
      if (!seqMatch) continue;
      const seq = seqMatch[1];

      // 날짜 (.date)
      // span.blind 제거 후 텍스트 추출
      $row.find(".date .blind").remove();
      const dateStr = $row.find(".date").text().trim();

      console.log(`[${i + 1}] Fetching detail for: ${title} (${dateStr})`);

      // 상세 페이지 크롤링
      const detailUrl = `${DETAIL_URL_BASE}${seq}`;
      const detailHtml = await fetchHtml(detailUrl);
      
      if (!detailHtml) {
          console.error(`Failed to fetch detail page for seq ${seq}`);
          continue;
      }
      
      const $detail = cheerio.load(detailHtml);

      // 본문 추출
      let $contentDiv = $detail(".board_view_con");
      if ($contentDiv.length === 0) {
          $contentDiv = $detail(".view_cont");
      }

      let contentHtml = "";
      const images: string[] = [];

      if ($contentDiv.length > 0) {
        // 불필요한 태그 제거 (script, style)
        $contentDiv.find("script, style").remove();

        contentHtml = $contentDiv.html() || "";

        // 이미지 링크 추출
        $contentDiv.find("img").each((_, img) => {
          let src = $(img).attr("src");
          if (src) {
            if (!src.startsWith("http")) {
              src = "https://www.gwangju.go.kr" + src;
            }
            images.push(src);
          }
        });
      }

      // 결과 객체 생성
      const item = {
        origin_id: `GWANGJU_${seq}`,
        source: "광주광역시청",
        title: title,
        content: contentHtml, // 전체 HTML 저장
        link: detailUrl,
        images: images,
        published_at: new Date(dateStr).toISOString(), // 날짜 변환
        status: "collected",
      };

      console.log(`[${i + 1}] Saving to DB: ${title}`);

      // Supabase 저장 (Upsert)
      const { error } = await supabase
        .from("press_releases")
        .upsert(item, { onConflict: "origin_id" });

      if (error) {
        console.error(`Failed to save ${title}:`, error);
        console.error(`Error Details:`, JSON.stringify(error, null, 2));
      } else {
        console.log(`Saved successfully!`);
      }

      // 서버 부하 방지 (1초 대기)
      await new Promise(resolve => setTimeout(resolve, 1000));

    } catch (error) {
      console.error(`Error processing row ${i}:`, error);
    }
  }
}

crawlGwangju();

File: scripts/crawl_gwangju.py
import requests
from bs4 import BeautifulSoup
import json
import re

# 광주광역시청 보도자료 URL
LIST_URL = "https://www.gwangju.go.kr/boardList.do?pageId=www789&boardId=BD_0000000027"
DETAIL_URL_BASE = "https://www.gwangju.go.kr/boardView.do?pageId=www789&boardId=BD_0000000027&seq="

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
}

def crawl_gwangju():
    print("Fetching list...")
    try:
        res = requests.get(LIST_URL, headers=HEADERS)
        res.raise_for_status()
    except Exception as e:
        print(f"Failed to fetch list: {e}")
        return

    soup = BeautifulSoup(res.text, "html.parser")
    
    # 목록 테이블 파싱 (구조 추정: table.board_list tbody tr)
    rows = soup.select("table.board_list tbody tr")
    if not rows:
        # 혹시 클래스명이 다를 경우를 대비해 일반적인 table row 선택
        rows = soup.select("table tbody tr")

    results = []

    print(f"Found {len(rows)} rows. Processing top 5...")

    for i, row in enumerate(rows[:5]):
        try:
            # 제목 셀 찾기 (보통 두 번째나 세 번째 td, class='subject'일 수도 있음)
            subject_td = row.select_one(".subject")
            if not subject_td:
                # 제목이 있는 td 찾기 (a 태그가 있는 곳)
                tds = row.select("td")
                for td in tds:
                    if td.find("a"):
                        subject_td = td
                        break
            
            if not subject_td:
                continue

            link = subject_td.find("a")
            title = link.get_text(strip=True)
            href = link.get("href")
            
            # href에서 seq 추출 (예: ?...&seq=12345...)
            seq_match = re.search(r'seq=(\d+)', href)
            if not seq_match:
                continue
            seq = seq_match.group(1)
            
            # 날짜 (보통 마지막이나 그 앞 td)
            date_td = row.select("td")[-2] # 조회수 앞이 날짜인 경우가 많음
            date = date_td.get_text(strip=True)

            print(f"[{i+1}] Fetching detail for: {title} ({date})")
            
            # 상세 페이지 크롤링
            detail_url = f"{DETAIL_URL_BASE}{seq}"
            detail_res = requests.get(detail_url, headers=HEADERS)
            detail_soup = BeautifulSoup(detail_res.text, "html.parser")

            # 본문 추출 (일반적인 보도자료 본문 클래스: board_view_con, view_cont 등)
            content_div = detail_soup.select_one(".board_view_con") or detail_soup.select_one(".view_cont")
            
            content_html = ""
            content_text = ""
            images = []

            if content_div:
                # 불필요한 태그 제거 (script, style)
                for script in content_div(["script", "style"]):
                    script.decompose()
                
                content_html = str(content_div)
                content_text = content_div.get_text("\n", strip=True)

                # 이미지 링크 추출
                for img in content_div.select("img"):
                    src = img.get("src")
                    if src:
                        if not src.startswith("http"):
                            src = "https://www.gwangju.go.kr" + src
                        images.append(src)

            # 결과 저장
            results.append({
                "title": title,
                "date": date,
                "seq": seq,
                "link": detail_url,
                "images": images,
                "content_preview": content_text[:100] + "..."
            })

        except Exception as e:
            print(f"Error processing row {i}: {e}")
            continue

    # 결과 출력
    print(json.dumps(results, indent=2, ensure_ascii=False))

if __name__ == "__main__":
    crawl_gwangju()
